File: 3312-sorted-gcd-pair-queries/3312-sorted-gcd-pair-queries.java
class Solution {
    public int[] gcdValues(int[] A, long[] queries) {
        int max = Arrays.stream(A).max().getAsInt();

        int[] freq = new int[max + 1];
        long[] GCD = new long[max + 1];

        for (int a : A) freq[a]++;

        for (int i = max; i > 0; i--) {
            long sum = 0, extra = 0;
            for (int j = i; j <= max; j += i) {
                sum += freq[j];
                extra += GCD[j];
            }
            GCD[i] = sum * (sum - 1) / 2 - extra;
        }

        Arrays.parallelPrefix(GCD, Long::sum);
        int n = queries.length;

        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            long q = queries[i];
            int l = 0, h = max + 1;
            while (l < h) {
                int m = (l + h) >>> 1;
                if (GCD[m] <= q) l = m + 1;
                else h = m;
            }
            res[i] = l;
        }

        return res;
    }
}

File: 3312-sorted-gcd-pair-queries/3312-sorted-gcd-pair-queries.cpp
class Solution {
    using ll = long long;
public:
    vector<int> gcdValues(vector<int>& A, vector<long long>& queries) {
        int mx = ranges::max(A);
        vector<int> freq(mx + 1, 0);
        vector<ll> GCD(mx + 1, 0);
        
        for (auto& a : A) freq[a]++;
        
        for (int i = mx; i > 0; i--) {
            ll sm = 0, extra = 0;
            for (int j = i; j <= mx; j += i)
                sm += freq[j], extra += GCD[j];
            GCD[i] = sm * (sm - 1) / 2 - extra;
        }
        
        partial_sum(GCD.begin(), GCD.end(), GCD.begin());
        
        vector<int> res(queries.size());
        for (int i = 0; i < queries.size(); i++)
            res[i] = ranges::upper_bound(GCD, queries[i]) - GCD.begin();
            
        return res;
    }
};